### ISSUE TITLE
Fleet in Kubernetes - Launcher Enrolls a Node but No Data Can Be Queried In Fleet

### ISSUE THREAD

**Issue by MattMencel** [kolide/fleet#1650]
### What version of `fleet` are you using (`fleet version --full`)?

image: kolide/fleet:1.0.5

### What operating system are you using?

MacOS 10.11.6

### What did you do?

Deployed Fleet to a self-hosted K8S cluster running in VMware.  Since I'm self-hosted, instead of using LoadBalancer in the fleet-loadbalancer service, I used NodePort.  It maps an external port (30674) on the k8s nodes to port 443 on the fleet-webserver pods.

Built the launcher tool on my Mac desktop.

Started launcher.

```
./build/launcher --hostname=kube-worker0:30674 --root_directory=$(mktemp -d) --enroll_secret=SECRET --insecure
{"build":"04e0bc47d67966b9aba66342f1e202696056617b","caller":"launcher.go:103","component":"launcher","level":"info","msg":"started kolide launcher","ts":"2017-11-22T21:42:22.371473397Z","version":"0.3.0-14-g04e0bc4"}
{"caller":"launcher.go:293","component":"launcher","grpc_secure":true,"level":"info","msg":"dialing grpc server","server":"kube-worker0.wiu.edu:30674","tls_secure":false,"ts":"2017-11-22T21:42:22.372241307Z"}
{"caller":"extension.go:358","component":"launcher","err":"sending status logs: writing buffered logs: writing logs: transport error sending logs: rpc error: code = Unknown desc = decoding status log: json: cannot unmarshal number into Go struct field StatusLog.s of type string","level":"info","ts":"2017-11-22T21:43:22.843015999Z"}
```

It enrolls the node in fleet and I can see it in the UI at https://kube-worker0:30674/hosts/manage

However Launcher on my desktop just repeats this error over and over

```
{"caller":"extension.go:358","component":"launcher","err":"sending status logs: writing buffered logs: writing logs: transport error sending logs: rpc error: code = Unknown desc = decoding status log: json: cannot unmarshal number into Go struct field StatusLog.s of type string","level":"info","ts":"2017-11-22T21:43:22.843015999Z"}
```
... and I never see any data appear in fleet that I can query about my node.

### What did you expect to see?

Expected to be able to query data about my newly enrolled Mac node.

### What did you see instead?

No queryable data.... and the error log from launcher.

I deployed fleet using the examples/kubernetes/fleet-*.yaml files.

The only changes made were... I put everything in namespace=fleet, and I changed the service to use NodePort and not LoadBalancer.

I can provide more debugging if you'd like.  Just let me know what you need.

**Comment by groob** [comment]
FWIW, I don't think this is related to `NodePort` vs `LoadBalancer` or k8s configuration at all. 

I think this was fixed in https://github.com/kolide/fleet/pull/1582 and the bug should not be there in master. The 1.0.6 release will resolve this. 
Related Launcher issue https://github.com/kolide/launcher/issues/183

You can also use the osquery binary instead of the launcher one if you want to wait for the next release.

**Comment by groob** [comment]
I pushed a new docker tag `kolide/fleet:63d60d5` which is a build from master. Would love to know if that works for you.

**Comment by groob** [comment]
Closing after debugging with @MattMencel in chat. It appears that the configuration used by flannel in the k8s cluster was preventing grpc connections. 

Used the Go script below to debug the healthcheck endpoint

```
package main

import (
	"context"
	"crypto/tls"
	"flag"
	"fmt"
	"log"
	"strings"

	pb "github.com/kolide/agent-api"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
)

func main() {
	var (
		flHostname = flag.String("hostname", "", "host:port")
		flTLS      = flag.Bool("tls", false, "use transport credentials")
	)
	flag.Parse()
	client := pb.NewApiClient(remoteConn(*flHostname, *flTLS))
	resp, err := client.CheckHealth(context.Background(), &pb.AgentApiRequest{})
	fmt.Println(resp.GetStatus())
	fmt.Println(resp, err)
}

func remoteConn(hostname string, useTLS bool) *grpc.ClientConn {
	var opts []grpc.DialOption
	if useTLS {
		justhost := strings.SplitN(hostname, ":", 2)[0]
		fmt.Printf("using tls with ServerName %s\n", justhost)
		opts = append(opts, grpc.WithTransportCredentials(
			credentials.NewTLS(&tls.Config{ServerName: justhost}),
		))
	} else {
		fmt.Println("not using TLS")
		opts = append(opts, grpc.WithInsecure())
	}
	conn, err := grpc.Dial(
		hostname,
		opts...,
	)
	if err != nil {
		log.Fatalf("did not connect: %v", err)
	}
	return conn
}
```

**Comment by frits-v** [comment]
@groob, I'm interested in the solution as I'm facing the same issue. I'm not entirely running the same setup though, using google hosted kubernetes in GKE. I'm not using flannel.

**Comment by groob** [comment]
@frots how are you exposing fleet? `type Loadbalancer` or Ingress?

Are you using the kolide launcher, or just plain osqueryd?

Did you run the health check script I posted above?

**Comment by frits-v** [comment]
Hi @groob,

I'm using `type LoadBalancer` and the kolide launcher.

The health check gives the following output:
```
using tls with ServerName
UNKNOWN
<nil> rpc error: code = Unavailable desc = all SubConns are in TransientFailure
```

Thanks.

Edit: I ran a few more tests today:
- The `osqueryd` binary does connect, it generates query result logs
- When tunneling the connection from the `launcher` binary directly to the fleet-webserver pod with `kube port-forward` it also doesn't work

**Comment by gjyoung1974** [comment]
I realize this is closed, I'm facing the same issue. @frots did you resolve this?
Thanks!

**Comment by frits-v** [comment]
@gjyoung1974 I did not resolve the issue, did not have the time to further
investigate.

**Comment by rowbutt** [comment]
tl;dr if you're terminating TLS w/ the fleet pod you need an L3/4 LB. it could be accomplished w/ a service `type: LoadBalancer` on GKE but I'm too cheap for that and just pinned a nodeport. 

service:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: fleet-nodeport
spec:
  ports:
  - name: https
    nodePort: 30443
    protocol: TCP
    port: 443
  selector:
    app: fleet
  type: NodePort
```

relevant workload bits:
```yaml
      containers:
      - name: fleet
        image: kolide/fleet:1.0.6
        command: ["fleet", "serve"]
        ports:
          - containerPort: 443
        env:
          - name: KOLIDE_SERVER_ADDRESS
            value: "0.0.0.0:443"
          - name: KOLIDE_SERVER_CERT
            value: "/secrets/fleet-tls/tls.crt"
          - name: KOLIDE_SERVER_KEY
            value: "/secrets/fleet-tls/tls.key"
        volumeMounts:
          - name: fleet-tls
            mountPath: "/secrets/fleet-tls"
            readOnly: true
      volumes:
        - name: fleet-tls
          secret:
            secretName: fleet-tls
```

here are the relevant bits from an enrolled node's plist on macOS (high sierra)
```xml
            <key>KOLIDE_LAUNCHER_ROOT_DIRECTORY</key>
            <string>/var/launcher/my.server.com-30443</string>
            <key>KOLIDE_LAUNCHER_HOSTNAME</key>
            <string>my.server.com:30443</string>
            <key>KOLIDE_LAUNCHER_ENROLL_SECRET_PATH</key>
            <string>/etc/launcher/secret</string>
            <key>KOLIDE_LAUNCHER_OSQUERYD_PATH</key>
            <string>/usr/local/launcher/bin/osqueryd</string>
            <key>KOLIDE_LAUNCHER_UPDATE_CHANNEL</key>
            <string>beta</string>
```

**Comment by frits-v** [comment]
Hi @j0npulsifer -- as mentioned before, I have configured a service `type: LoadBalancer` on GKE.

edit:
Never mind, I realized a new release was published after my issue 1.0.6 *does* solve the problem for me.

Thanks @groob
